components/webgpu/identity.rs

```diff
@@ -0,0 +1,115 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use ipc_channel::ipc::IpcSender;
+use serde::{Deserialize, Serialize};
+use wgpu::{
+    hub::{GlobalIdentityHandlerFactory, IdentityHandler, IdentityHandlerFactory},
+    id::{
+        AdapterId, BindGroupId, BindGroupLayoutId, BufferId, CommandBufferId, ComputePipelineId,
+        DeviceId, PipelineLayoutId, RenderPipelineId, SamplerId, ShaderModuleId, SurfaceId,
+        SwapChainId, TextureId, TextureViewId, TypedId,
+    },
+};
+use wgt::Backend;
+
+#[derive(Clone, Copy, Debug, Deserialize, Serialize)]
+pub enum WebGPUMsg {
+    FreeAdapter(AdapterId),
+    FreeDevice(DeviceId),
+    FreeBuffer(BufferId),
+    FreeSwapChain(SwapChainId),
+    FreePipelineLayout(PipelineLayoutId),
+    FreeComputePipeline(ComputePipelineId),
+    FreeRenderPipeline(RenderPipelineId),
+    FreeBindGroup(BindGroupId),
+    FreeBindGroupLayout(BindGroupLayoutId),
+    FreeCommandBuffer(CommandBufferId),
+    FreeTexture(TextureId),
+    FreeTextureView(TextureViewId),
+    FreeSampler(SamplerId),
+    FreeSurface(SurfaceId),
+    FreeShaderModule(ShaderModuleId),
+}
+
+#[derive(Debug)]
+pub struct IdentityRecycler<I> {
```

Comment:
we technically don't have to make this generic over `I` even. It's not exposed to the outside, and it's used behind the strongly typed associated types.